### PR TITLE
zephyr-sdk: init at 0.16.1

### PR DESCRIPTION
**Pull request by mibmo**
Based on #141694

## Description of changes

Adds the zephyr-sdk package

## Things done

Should be the same as #141694 with the exception of this being a newer version. Nonetheless testing is probably a good idea. I don't have access to any zephyr compatible boards right now, so if anyone can test it would be greatly appreciated!

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by otavio — COMMENTED** [review]
How are you testing this?

**Comment by mibmo** [comment]
> How are you testing this?

Just building it - it's just a collection of files for the SDK to work. But as the PR says, testing will have to be done with someone access to a Zephyr compatible board - I'm not around my workstation right now and won't be for the foreseeable future, so I can't do the actual testing (although I doubt it doesn't work since this should be identical to what Zephyr ships :))

If you look at the previous PR https://github.com/NixOS/nixpkgs/pull/141694 you can see how they tested it - ideally we'd do that here too.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please set `meta.sourceProvenance`.

**Comment by mibmo** [comment]
> sourceProvenance

I'm reading [nixpkgs manual section meta/sourceProvenance](https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance) and as far as I can tell this would be `lib.sourceTypes.binaryFirmware`. Does that sound right? :)

**Comment by mibmo** [comment]
@dotlambda I've set sourceProvenance :) lmk if it's not the right type.

File: pkgs/development/embedded/zephyr-sdk/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    homepage = "https://www.zephyrproject.org/";
+    description = "Collection of compilers and tools for the Zephyr RTOS";
+    license = licenses.asl20;
+    platforms = [ "x86_64-linux" "aarch64-linux" ];
+    maintainers = with maintainers; [ mib ];
+    sourceProvenance = with sourceTypes; [ binaryFirmware ];
```
[the remark this answers is not in the document]
Comment (mibmo) [reply pkgs/development/embedded/zephyr-sdk/default.nix:37]: Gotcha, thanks! Was struggling a bit to understand it, thanks. :)

File: pkgs/development/embedded/zephyr-sdk/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{ lib
+, stdenv
+, fetchurl
+}:
+stdenv.mkDerivation (finalAttrs: {
```
[the remark this answers is not in the document]
Comment (mibmo) [reply pkgs/development/embedded/zephyr-sdk/default.nix:5]: I'll add it; seems like a lot of packages have it, but am curious as to why? Do you happen to know? :)

File: pkgs/development/embedded/zephyr-sdk/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+    cp -r . $out
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    homepage = "https://www.zephyrproject.org/";
```
[the remark this answers is not in the document]
Comment (mibmo) [reply pkgs/development/embedded/zephyr-sdk/default.nix:32]: Sgtm. I think it's more accurate. :)

File: pkgs/development/embedded/zephyr-sdk/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (dotlambda) [reply pkgs/development/embedded/zephyr-sdk/default.nix:6]: It's just a convention.

[the remark this answers is not in the document]
Comment (mibmo) [reply pkgs/development/embedded/zephyr-sdk/default.nix:6]: Fair enough! :) Added it to the commit, thanks for letting me know.

**Comment by mibmo** [comment]
I don't know if this is a bad idea, but since this derivation only really downloads and then unpacks something (and it takes quite a while to unpack, at least on my machine ...), is it worth it to exclude this from builds somehow?

**Comment by rndmono** [comment]
Hi @mibmo, thank you for for your work on this! I'd really love to be able to build Zephyr-applications on NixOS.
I did a checkout of this PR (a34b83d064ea06d5100df50cd3f6d4a0acce2017), and could build the package. But when running e.g. aarch64-zephyr-elf-gcc, I'd get a segmentation fault, and `strace` would show the following:
```
> strace ./arm-zephyr-eabi-gcc-12.2.0
execve("./arm-zephyr-eabi-gcc-12.2.0", ["./arm-zephyr-eabi-gcc-12.2.0"], 0x7ffee560fce0 /* 80 vars */) = -1 ENOENT (No such file or directory)
strace: exec: No such file or directory
+++ exited with 1 +++
```
I think the ELF files require some patching, since they expect an interpreter in `/lib64/`:
```
> file ./arm-zephyr-eabi-gcc-12.2.0
./arm-zephyr-eabi-gcc-12.2.0: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=682c2df417b995c9409e352a15d59632f76b70ed, stripped
```
Fortunately, I was able to get things running by applying the following:
```
diff --git a/pkgs/development/embedded/zephyr-sdk/default.nix b/pkgs/development/embedded/zephyr-sdk/default.nix
index 60a04c5c3ce1..2eac852b01de 100644
--- a/pkgs/development/embedded/zephyr-sdk/default.nix
+++ b/pkgs/development/embedded/zephyr-sdk/default.nix
@@ -1,12 +1,20 @@
 { lib
 , stdenv
 , fetchurl
+, python38
+, autoPatchelfHook
 }:
 
 stdenv.mkDerivation (finalAttrs: {
   pname = "zephyr-sdk";
   version = "0.16.1";
 
+  nativeBuildInputs = [
+    autoPatchelfHook
+    stdenv.cc.cc
+    python38
+  ];
+
   src = let
     inherit (finalAttrs) version;
   in builtins.getAttr stdenv.hostPlatform.system {
```
(`lib`)`python38` and `stdenv.cc.cc` seem to be required for the ZehyrSDK, without them (after adding the `autoPatchelfHook`) the nix would fail to build the package.

With these changes, I could successfully build a Zephyr sample for an nRF-board, flash it to the board, and it worked!

**Comment by mibmo** [comment]
Thank you so much for your testing, @rndmono! I still don't have a board available to test sadly, but I'll implement your changes :) Very glad to hear you got it working though!

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by prusnak** [comment]
Can we upgrade this PR to latest ([0.16.4](https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.4)) and maybe also enable macOS/Darwin build? Tarballs are available from the release page.

**Comment by mibmo** [comment]
Closing in favor of #294092, as either it's going to progress _or zephyr-sdk won't make its way into nixpkgs._ (at least how i see it)

**Pull request closed without merging**